File: Nudge-Neurodivergent-task-tracker/PomodoroTimer.java
import java.time.LocalDateTime;
import java.time.Duration;

/**
 * Manages Pomodoro-style work sessions with breaks and time management
 */
public class PomodoroTimer {
    public enum SessionType {
        WORK("🍅 Work Session"),
        SHORT_BREAK("☕ Short Break"),
        LONG_BREAK("🌿 Long Break");
        
        private final String displayName;
        
        SessionType(String displayName) {
            this.displayName = displayName;
        }
        
        public String getDisplayName() { return displayName; }
    }
    
    public enum TimerState {
        IDLE, RUNNING, PAUSED, COMPLETED
    }
    
    private int workDuration; // minutes
    private int shortBreakDuration; // minutes
    private int longBreakDuration; // minutes
    private int sessionsUntilLongBreak;
    
    private SessionType currentSessionType;
    private TimerState state;
    private LocalDateTime sessionStartTime;
    private int currentDuration;
    private int completedSessions;
    private int totalFocusTime; // total minutes
    
    private Task currentTask;
    
    public PomodoroTimer() {
        this.workDuration = 25;
        this.shortBreakDuration = 5;
        this.longBreakDuration = 15;
        this.sessionsUntilLongBreak = 4;
        
        this.currentSessionType = SessionType.WORK;
        this.state = TimerState.IDLE;
        this.completedSessions = 0;
        this.totalFocusTime = 0;
    }
    
    public void startSession(Task task) {
        this.currentTask = task;
        this.sessionStartTime = LocalDateTime.now();
        this.state = TimerState.RUNNING;
        
        switch (currentSessionType) {
            case WORK:
                currentDuration = workDuration;
                break;
            case SHORT_BREAK:
                currentDuration = shortBreakDuration;
                break;
            case LONG_BREAK:
                currentDuration = longBreakDuration;
                break;
        }
        
        System.out.println("\n🍅 " + currentSessionType.getDisplayName() + " started!");
        if (task != null && currentSessionType == SessionType.WORK) {
            System.out.println("📋 Working on: " + task.getName());
        }
        System.out.println("⏰ Duration: " + currentDuration + " minutes");
        System.out.println("💡 Focus tip: " + getFocusTip());
    }
    
    public void pauseSession() {
        if (state == TimerState.RUNNING) {
            state = TimerState.PAUSED;
            System.out.println("⏸️ Session paused. Take a moment to breathe.");
        }
    }
    
    public void resumeSession() {
        if (state == TimerState.PAUSED) {
            state = TimerState.RUNNING;
            System.out.println("▶️ Session resumed. You've got this!");
        }
    }
    
    public void completeSession() {
        if (state != TimerState.RUNNING && state != TimerState.PAUSED) {
            return;
        }
        
        state = TimerState.COMPLETED;
        completedSessions++;
        
        if (currentSessionType == SessionType.WORK) {
            totalFocusTime += workDuration;
            System.out.println("\n🎉 Work session completed!");
            System.out.println("✨ " + getCompletionEncouragement());
            
            // Determine next session type
            if (completedSessions % sessionsUntilLongBreak == 0) {
                currentSessionType = SessionType.LONG_BREAK;
                System.out.println("🌿 Time for a long break! You've earned it!");
            } else {
                currentSessionType = SessionType.SHORT_BREAK;
                System.out.println("☕ Time for a short break!");
            }
        } else {
            System.out.println("\n🔄 Break completed! Ready to focus again?");
            currentSessionType = SessionType.WORK;
        }
        
        state = TimerState.IDLE;
    }
    
    public void skipSession() {
        System.out.println("⏭️ Session skipped. That's okay - listen to your brain!");
        state = TimerState.IDLE;
        
        if (currentSessionType != SessionType.WORK) {
            currentSessionType = SessionType.WORK;
        }
    }
    
    public String getTimeRemaining() {
        if (state != TimerState.RUNNING) {
            return "Timer not running";
        }
        
        Duration elapsed = Duration.between(sessionStartTime, LocalDateTime.now());
        long elapsedMinutes = elapsed.toMinutes();
        long remaining = Math.max(0, currentDuration - elapsedMinutes);
        
        return remaining + " minutes remaining";
    }
    
    public boolean isSessionComplete() {
        if (state != TimerState.RUNNING) {
            return false;
        }
        
        Duration elapsed = Duration.between(sessionStartTime, LocalDateTime.now());
        return elapsed.toMinutes() >= currentDuration;
    }
    
    private String getFocusTip() {
        String[] workTips = {
            "Put your phone in another room",
            "Close unnecessary browser tabs",
            "Take three deep breaths before starting",
            "Set your intention for this session",
            "Remember: progress over perfection",
            "One task at a time is plenty"
        };
        
        String[] breakTips = {
            "Step away from your workspace",
            "Do some gentle stretching",
            "Look at something far away",
            "Drink some water",
            "Take deep breaths",
            "Move your body a little"
        };
        
        String[] tips = (currentSessionType == SessionType.WORK) ? workTips : breakTips;
        return tips[(int) (Math.random() * tips.length)];
    }
    
    private String getCompletionEncouragement() {
        String[] encouragements = {
            "Your brain did amazing work!",
            "Look at you staying focused!",
            "That's the kind of energy we love!",
            "You're building such great habits!",
            "Your future self is so grateful!",
            "Progress is progress - celebrate it!"
        };
        
        return encouragements[(int) (Math.random() * encouragements.length)];
    }
    
    public void displayStats() {
        System.out.println("\n🍅 POMODORO STATS");
        System.out.println("-".repeat(30));
        System.out.printf("📊 Completed Sessions: %d\n", completedSessions);
        System.out.printf("⏰ Total Focus Time: %d minutes (%.1f hours)\n", 
                         totalFocusTime, totalFocusTime / 60.0);
        System.out.printf("🎯 Current Session: %s\n", currentSessionType.getDisplayName());
        System.out.printf("📱 Status: %s\n", state);
        
        if (state == TimerState.RUNNING) {
            System.out.printf("⏳ %s\n", getTimeRemaining());
        }
    }
    
    public void customizeSettings(int work, int shortBreak, int longBreak, int sessionsForLong) {
        this.workDuration = work;
        this.shortBreakDuration = shortBreak;
        this.longBreakDuration = longBreak;
        this.sessionsUntilLongBreak = sessionsForLong;
        
        System.out.println("⚙️ Pomodoro settings updated!");
        System.out.printf("   Work: %d min | Short Break: %d min | Long Break: %d min\n",
                         work, shortBreak, longBreak);
    }
    
    // Getters
    public SessionType getCurrentSessionType() { return currentSessionType; }
    public TimerState getState() { return state; }
    public int getCompletedSessions() { return completedSessions; }
    public int getTotalFocusTime() { return totalFocusTime; }
    public Task getCurrentTask() { return currentTask; }
    public int getWorkDuration() { return workDuration; }
    public int getShortBreakDuration() { return shortBreakDuration; }
    public int getLongBreakDuration() { return longBreakDuration; }
}

File: Nudge-Neurodivergent-task-tracker/pomodoro_timer.py
"""
Pomodoro timer for focus sessions
"""

from datetime import datetime, timedelta
from enum import Enum
from typing import Optional
from dataclasses import dataclass


class TimerState(Enum):
    """Timer states"""
    IDLE = "idle"
    RUNNING = "running"
    PAUSED = "paused"
    BREAK = "break"


class SessionType(Enum):
    """Types of Pomodoro sessions"""
    WORK = "Work"
    SHORT_BREAK = "Short Break"
    LONG_BREAK = "Long Break"


@dataclass
class PomodoroSession:
    """A completed Pomodoro session"""
    start_time: datetime
    end_time: datetime
    session_type: SessionType
    task_name: Optional[str] = None
    completed: bool = True


class PomodoroTimer:
    """Pomodoro timer with customizable intervals"""
    
    def __init__(self):
        self.work_duration = 25  # minutes
        self.short_break_duration = 5  # minutes
        self.long_break_duration = 15  # minutes
        self.sessions_until_long_break = 4
        
        self.current_session_start: Optional[datetime] = None
        self.current_session_type = SessionType.WORK
        self.current_task_name: Optional[str] = None
        self.state = TimerState.IDLE
        self.paused_time: Optional[datetime] = None
        self.elapsed_when_paused = timedelta()
        
        self.completed_sessions: list[PomodoroSession] = []
        self.work_sessions_today = 0
    
    def start_session(self, task_name: Optional[str] = None):
        """Start a new Pomodoro session"""
        if self.state != TimerState.IDLE:
            print("🍅 Session already in progress!")
            return
        
        self.current_session_start = datetime.now()
        self.current_task_name = task_name
        self.state = TimerState.RUNNING
        self.elapsed_when_paused = timedelta()
        
        duration = self._get_current_duration()
        session_name = self.current_session_type.value
        
        if task_name:
            print(f"🍅 Started {session_name} session for: {task_name}")
        else:
            print(f"🍅 Started {session_name} session")
        
        print(f"⏰ Duration: {duration} minutes")
        print("💡 Use menu options to pause/complete/skip when ready")
    
    def pause_session(self):
        """Pause the current session"""
        if self.state != TimerState.RUNNING:
            print("🌸 No active session to pause")
            return
        
        self.paused_time = datetime.now()
        elapsed = self.paused_time - self.current_session_start
        self.elapsed_when_paused = elapsed
        self.state = TimerState.PAUSED
        print("⏸️ Session paused")
    
    def resume_session(self):
        """Resume a paused session"""
        if self.state != TimerState.PAUSED:
            print("🌸 No paused session to resume")
            return
        
        # Adjust start time to account for pause duration
        pause_duration = datetime.now() - self.paused_time
        self.current_session_start += pause_duration
        self.state = TimerState.RUNNING
        print("▶️ Session resumed")
    
    def complete_session(self):
        """Complete the current session"""
        if self.state not in [TimerState.RUNNING, TimerState.PAUSED]:
            print("🌸 No active session to complete")
            return
        
        end_time = datetime.now()
        session = PomodoroSession(
            start_time=self.current_session_start,
            end_time=end_time,
            session_type=self.current_session_type,
            task_name=self.current_task_name,
            completed=True
        )
        
        self.completed_sessions.append(session)
        
        if self.current_session_type == SessionType.WORK:
            self.work_sessions_today += 1
            print(f"✅ Work session completed! That's {self.work_sessions_today} today!")
            
            # Suggest next session type
            if self.work_sessions_today % self.sessions_until_long_break == 0:
                self.current_session_type = SessionType.LONG_BREAK
                print(f"🎉 Time for a {self.long_break_duration}-minute long break!")
            else:
                self.current_session_type = SessionType.SHORT_BREAK
                print(f"☕ Time for a {self.short_break_duration}-minute short break!")
        else:
            print("✅ Break completed! Ready for the next work session?")
            self.current_session_type = SessionType.WORK
        
        self._reset_session()
    
    def skip_session(self):
        """Skip the current session"""
        if self.state not in [TimerState.RUNNING, TimerState.PAUSED]:
            print("🌸 No active session to skip")
            return
        
        print("⏭️ Session skipped - that's totally okay!")
        self._reset_session()
    
    def _reset_session(self):
        """Reset session state"""
        self.current_session_start = None
        self.current_task_name = None
        self.state = TimerState.IDLE
        self.paused_time = None
        self.elapsed_when_paused = timedelta()
    
    def _get_current_duration(self) -> int:
        """Get duration for current session type"""
        if self.current_session_type == SessionType.WORK:
            return self.work_duration
        elif self.current_session_type == SessionType.SHORT_BREAK:
            return self.short_break_duration
        else:
            return self.long_break_duration
    
    def get_time_remaining(self) -> str:
        """Get remaining time in current session"""
        if self.state == TimerState.IDLE:
            return "No active session"
        
        if self.state == TimerState.PAUSED:
            elapsed = self.elapsed_when_paused
        else:
            elapsed = datetime.now() - self.current_session_start
        
        total_duration = timedelta(minutes=self._get_current_duration())
        remaining = total_duration - elapsed
        
        if remaining.total_seconds() <= 0:
            return "Time's up! ⏰"
        
        minutes = int(remaining.total_seconds() // 60)
        seconds = int(remaining.total_seconds() % 60)
        return f"{minutes:02d}:{seconds:02d}"
    
    def customize_settings(self, work_minutes: int, short_break_minutes: int, 
                          long_break_minutes: int, sessions_for_long: int):
        """Customize timer settings"""
        self.work_duration = work_minutes
        self.short_break_duration = short_break_minutes
        self.long_break_duration = long_break_minutes
        self.sessions_until_long_break = sessions_for_long
        
        print("⚙️ Pomodoro settings updated!")
        print(f"   Work: {work_minutes}min")
        print(f"   Short break: {short_break_minutes}min")
        print(f"   Long break: {long_break_minutes}min")
        print(f"   Sessions until long break: {sessions_for_long}")
    
    def display_stats(self):
        """Display Pomodoro statistics"""
        print("\n🍅 POMODORO STATS")
        print("-" * 30)
        print(f"📊 Total sessions: {len(self.completed_sessions)}")
        print(f"💼 Work sessions today: {self.work_sessions_today}")
        
        total_work_time = sum(
            (session.end_time - session.start_time).total_seconds() / 60
            for session in self.completed_sessions
            if session.session_type == SessionType.WORK
        )
        print(f"⏱️ Total focus time: {total_work_time:.0f} minutes ({total_work_time/60:.1f} hours)")
        
        if self.state != TimerState.IDLE:
            print(f"🔄 Current: {self.current_session_type.value} - {self.get_time_remaining()}")
    
    def get_state(self) -> TimerState:
        return self.state
    
    def get_completed_sessions(self) -> int:
        return len([s for s in self.completed_sessions if s.session_type == SessionType.WORK])
    
    def get_total_focus_time(self) -> float:
        """Get total focus time in minutes"""
        return sum(
            (session.end_time - session.start_time).total_seconds() / 60
            for session in self.completed_sessions
            if session.session_type == SessionType.WORK
        )